--- scripting.c.orig	Mon Jun 16 07:45:11 2003
+++ scripting.c	Sat Jun 21 10:54:03 2003
@@ -404,6 +404,16 @@
 	c->return_val.u.aval->vals[i].type = v_void;
 }
 
+static void bSizeOf(Context *c) {
+    if ( c->a.argc!=2 )
+	error( c, "Wrong number of arguments" );
+    if ( c->a.vals[1].type!=v_arr && c->a.vals[1].type!=v_arrfree))
+	error( c, "Expected array argument" );
+
+    c->return_val.type = v_int;
+    c->return_val.u.ival = c->a.vals[1].aval->argc;
+}
+
 static void bStrlen(Context *c) {
 
     if ( c->a.argc!=2 )
@@ -2676,6 +2686,7 @@
     { "AskUser", bAskUser, 1 },
     { "PostNotice", bPostNotice, 1 },
     { "Array", bArray, 1 },
+    { "SizeOf", bSizeOf, 1 },
     { "Strsub", bStrsub, 1 },
     { "Strlen", bStrlen, 1 },
     { "Strstr", bStrstr, 1 },
